src/types/bytearray.rs

```diff
@@ -23,36 +23,55 @@ impl PyByteArray {
 
     /// Creates a new Python `bytearray` object with an `init` closure to write its contents.
     /// Before calling `init` the bytearray is zero-initialised.
+    /// * If `init` returns `Err(e)`, `new_with` will return `Err(e)`.
+    /// * If `init` returns `Ok(new_len)`, the allocated bytearray will be truncated to length
+    ///   `new_len` and `new_with` will return `Ok(&PyByteArray)`.
+    /// * If `init` returns `Ok(None)`, `new_with` will return `Ok(&PyByteArray)`.
     ///
-    /// Panics if out of memory.
+    /// Iff Python raises a MemoryError on the allocation, `new_with` will return it inside `Err`.
     ///
     /// # Example
     /// ```
     /// use pyo3::{prelude::*, types::PyByteArray};
-    /// Python::with_gil(|py| {
+    /// Python::with_gil(|py| -> PyResult<()> {
     ///     let py_bytearray = PyByteArray::new_with(py, 10, |bytes: &mut [u8]| {
     ///         bytes.copy_from_slice(b"Hello Rust");
-    ///     });
+    ///         Ok(None)
+    ///     })?;
     ///     let bytearray: &[u8] = unsafe { py_bytearray.as_bytes() };
     ///     assert_eq!(bytearray, b"Hello Rust");
+    ///     Ok(())
     /// });
     /// ```
-    pub fn new_with<F>(py: Python, len: usize, init: F) -> &PyByteArray
+    pub fn new_with<F>(py: Python, len: usize, init: F) -> PyResult<&PyByteArray>
     where
-        F: FnOnce(&mut [u8]),
+        F: FnOnce(&mut [u8]) -> PyResult<Option<usize>>,
     {
         unsafe {
-            let length = len as ffi::Py_ssize_t;
-            let pyptr = ffi::PyByteArray_FromStringAndSize(std::ptr::null(), length);
-            // Iff pyptr is null, py.from_owned_ptr(pyptr) will panic
-            let pybytearray = py.from_owned_ptr(pyptr);
+            let pyptr =
+                ffi::PyByteArray_FromStringAndSize(std::ptr::null(), len as ffi::Py_ssize_t);
+            if pyptr.is_null() {
+                return Err(PyErr::fetch(py));
+            }
             let buffer = ffi::PyByteArray_AsString(pyptr) as *mut u8;
             debug_assert!(!buffer.is_null());
             // Zero-initialise the uninitialised bytearray
             std::ptr::write_bytes(buffer, 0u8, len);
             // (Further) Initialise the bytearray in init
-            init(std::slice::from_raw_parts_mut(buffer, len));
-            pybytearray
+            match init(std::slice::from_raw_parts_mut(buffer, len)) {
+                Ok(Some(new_len)) if new_len < len => {
+                    if ffi::PyByteArray_Resize(pyptr, new_len as ffi::Py_ssize_t) != 0 {
+                        return Err(PyErr::fetch(py));
+                    }
+                }
+                Ok(_) => (),
+                Err(e) => {
+                    // Deallocate pyptr to avoid leaking the bytearray
+                    ffi::Py_DECREF(pyptr);
+                    return Err(e);
+                }
```

Comment:
An interesting thought about the `UninitializedPyBytes` API is that it could have a `Drop` impl which automatically deallocates the pointer, which would mean that it'd happen automatically on `Err` return without us having to do anything.